lrc/Pos/Lrc/Fts.hs

```diff
@@ -1,52 +1,250 @@
+    , (LocalSlotIndex 1, CoinIndex 327)
+    , ... ]
+@
+
+Since they are zipped together with their slot indices, we can restore the
+original ordering later (using 'arrangeBySlots').
+
+Now that selected coin indices and stakeholder coin index ranges are both
+sorted, we can walk through them in one pass! To do that, we 'traverse' the
+list of selected coin indices. As we do that, we monadically iterate through
+the table of balances, storing current stakeholder identifier and coin index
+range in a state (see 'FtsState').
+
+Let's take a look at the intermediate states during execution:
+
+1.  Initialize the state as @StakeholderId=aaa, CoinRange=[0..199]@.
+
+2.  Select the next coin, @LocalSlotIndex=0, CoinIndex=5@.
+
+3.  @CoinIndex@ falls within the @CoinRange@, yield slot leader
+    @LocalSlotIndex=0, StakeholderId=aaa@.
+
+5.  Select the next coin, @LocalSlotIndex=2, CoinIndex=61@.
+
+6.  @CoinInde@x falls within the @CoinRange@, yield slot leader
+    @LocalSlotIndex=2, StakeholderId=aa@a.
+
+7.  Select the next coin, @LocalSlotIndex=k, CoinIndex=205@.
+
+8.  @CoinIndex@ does not fall within the @CoinRange@.
+    Update the state to @StakeholderId=bbb, CoinRange=[200..249]@.
+
+9.  @CoinIndex@ falls within the @CoinRange@, yield slot leader
+    @LocalSlotIndex=k, StakeholderId=bbb@.
+
+10. Select the next coin, @LocalSlotIndex=1, CoinIndex=327@.
+
+11. @CoinIndex@ does not fall within the @CoinRange@.
+    Update the state to...
+
+Eventually, we end up with a list of slot leaders:
+
+@
+    [ (LocalSlotIndex 0, StakeholderId aaa)
+    , (LocalSlotIndex 2, StakeholderId aaa)
+    , (LocalSlotIndex k, StakeholderId bbb)
+    , (LocalSlotIndex 1, StakeholderId ccc)
+    , ... ]
+@
+
+and we're done. One thing to notice is that 'MonadIterator' provides us with a
+balance - the amount of coins a stakeholder has. So in order to maintain coin
+range, we have to accumulate (by summation) those balances - this way we get
+upper bounds. There's no need to calculate lower bounds, since every lower
+bound is equal to the upper bound of the previous stakeholder plus 1, and we
+transition to next coin only when the current coin index exceeds the upper
+bound.  Since coin indices are sorted, whenever we transition to the next coin,
+its index bona fide exceeds the previous upper bound (and thus it's more or
+equal to the current lower bound).
+
+-}
 followTheSatoshiM
     :: forall m . MonadIterator (StakeholderId, Coin) m
-    => SharedSeed -> Coin -> m (NonEmpty StakeholderId)
+    => SharedSeed
+    -> Coin
+    -> m SlotLeaders
 followTheSatoshiM _ totalCoins
     | totalCoins == mkCoin 0 = error "followTheSatoshiM: nobody has any stake"
 followTheSatoshiM (SharedSeed seed) totalCoins = do
-    res <- findLeaders (sortOn fst $ zip coinIndices [1..]) (mkCoin 0)
-    pure . fromList . map fst . sortOn snd $ res
+    ftsState <- ftsStateInit <$> nextStakeholder
+    res <-
+        flip evalStateT ftsState $
```

Comment:
If you don't like `flip evalStateT` there's [`evaluatingStateT`](https://github.com/serokell/universum/blob/70a9416b88624a388557b29a827182347d8654da/src/Monad/Trans.hs#L64) (about which you probably don't know yet) in `universum`